drivers/console/aarch64/console.S

```diff
@@ -3,103 +3,240 @@
+
+#if !ERROR_DEPRECATED
+	/* Set up compatiblity symbols for platforms using old console API. */
+	declare_console core
+
+	.globl	console_init
+	/* -----------------------------------------------
 	 * int console_init(uintptr_t base_addr,
 	 * unsigned int uart_clk, unsigned int baud_rate)
-	 * Function to initialize the console without a
-	 * C Runtime to print debug information. It saves
-	 * the console base to the data section.
+	 * Deprecated legacy function for initializing a
+	 * single console. Current platforms should call
+	 * console_<driver>_register for the relevant
+	 * driver instead. This needs to clear the console
+	 * list first because some legacy platforms called
+	 * console_init more than once and expect it to
+	 * overwrite previous console state.
 	 * In: x0 - console base address
 	 *     w1 - Uart clock in Hz
 	 *     w2 - Baud rate
-	 * out: return 1 on success else 0 on error
-	 * Clobber list : x1 - x4
+	 * Clobber list : x0 - x4, x14, x15
 	 * -----------------------------------------------
 	 */
 func console_init
-	/* Check the input base address */
-	cbz	x0, init_fail
-	adrp	x3, console_base
-	str	x0, [x3, :lo12:console_base]
-	b	console_core_init
-init_fail:
-	ret
+	adrp	x14, console_list
+	str	xzr, [x14, :lo12:console_list]
+	b	console_core_register
 endfunc console_init
+#endif
+
+	/* -----------------------------------------------
+	 * console_info_t *console_unregister(
+	 *                        console_info_t *console)
+	 * Function to find a specific console in the list
+	 * of currently active consoles and remove it.
+	 * Should usually only be called by
+	 * declare_console expansions.
+	 * In: x0 - address of console structure to remove
+	 * Out: x0 - removed address, or NULL if not found
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_unregister
+	adrp	x14, console_list
+	add	x14, x14, :lo12:console_list	/* X14 = ptr to first struct */
+	ldr	x1, [x14]			/* X3 = first struct */
+
+unregister_loop:
+	cbz	x1, unregister_not_found
+	cmp	x0, x1
+	b.eq	unregister_found
+	ldr	x14, [x14]			/* X14 = next ptr of struct */
+	ldr	x1, [x14]			/* X1 = next struct */
+	b	unregister_loop
+
+unregister_found:
+	ldr	x1, [x1]			/* X1 = next struct */
+	str	x1, [x14]			/* prev->next = cur->next */
+	ret
+
+unregister_not_found:
+	mov	x0, #0				/* return NULL if not found */
+	ret
+endfunc console_unregister
 
 	/* -----------------------------------------------
 	 * void console_uninit(void)
-	 * Function to finish the use of console driver.
-	 * It sets the console_base as NULL so that any
-	 * further invocation of `console_putc` or
-	 * `console_getc` APIs would return error.
+	 * Function to switch from boot to runtime state.
+	 * The name is misleading for compatibility with
+	 * the legacy API -- this used to uninitialize all
+	 * consoles, but now it just quiesces consoles
+	 * that are not marked as "runtime".
+	 * Clobber list: x0, x1
```

Comment:
> Also, platforms that do override plat_crash_console_init will need to remember to call console_start_crash to keep things working.

My intention is that platforms which do override plat_crash_console_init (I assume this would mostly be done temporarily to debug early boot failures) would also override plat_crash_console_putc and plat_crash_console_flush, so they wouldn't rely on the common infrastructure at all and don't need to call console_crash_start.

Also, getting plat_crash_console_init right for this case becomes much harder when your console is no longer just a simple UART with hardcoded initialization parameters. For example, the base address for the memory console I want to add is passed into BL31 via a parameter from BL2 (coreboot), which points to a memory structure that needs to be parsed at boot time because it is no longer accessible after MMU initialization. It would be hard to find that value again during a crash (especially because plat_crash_console_init does not permit a C environment, so you would force all console discovery to be written in assembly with a restricted register set). It's possible to store this in global variables in platform code, sure, but isn't it much easier if the console infrastructure does this for you?

> doesn't help the platforms I mentioned before, as it's only possible to have one serial console registered at a time

If this is the main point of contention I'd rather expand the system to support that as well, instead of implementing something that I think makes the use case I'm trying to solve more painful. Would you be okay with that?